KAFKA-7043: Modified plugin isolation whitelist with recently added converters (KIP-305)

rhauch opened the pull request:
Several recently-added converters are included in the plugin isolation whitelist, similarly to the `StringConverter`. This is a change in the implementation, and does not affect the approved KIP. Several unit tests were added to verify they are being loaded in isolation, again similarly to `StringConverter`.

These changes should be applied only to `trunk` and `2.0`, since these converters were added as part of KIP-305 for AK 2.0.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-7043: Modified plugin isolation whitelist with recently added converters

Several recently-added converters are included in the plugin isolation whitelist, similarly to the `StringConverter`.

rhauch commented:
@kkonstantine, @mageshn: would you please review this change? Thanks!

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java:
```diff
@@ -128,6 +128,12 @@
             + "|file\\..*"
             + "|converters\\..*"
             + "|storage\\.StringConverter"
+            + "|storage\\.SimpleHeaderConverter"
```
nit: Let's add it in the end of this list, in the order that appears in the unit test which also happens to be the lexicographic order (also a header converter and not a regular converter).

kkonstantine reviewed (approved):
LGTM with a minor suggestion. Unfortunately with several interfaces and abstract classes currently explicit listing is better than inverted regex.

rhauch committed:
KAFKA-7043: Reordered the whitelist lines to be lexicographical for readability

mageshn reviewed (approved):
LGTM. Thanks, @rhauch.

rhauch committed:
KAFKA-7043: Removed the SimpleHeaderConverter and will re-add as a separate PR

The SimpleHeaderConverter change needs to be backported, so this will be re-added as KAFKA-7047.

rhauch commented:
@kkonstantine, thanks for the suggestion. All of the converters are now in lexicographical order, and I've removed the header converter as I've logged a separate issue ([KAFKA-7047](https://issues.apache.org/jira/browse/KAFKA-7047)) for that so it can be backported to 1.1 as well (the numeric converters are trunk-only). Tests also use this same order.

rhauch commented:
Retest this please

rhauch commented:
retest this please.

rhauch commented:
The earlier test failures are unrelated to these changes, but requesting new builds.

rhauch commented:
@ewencp would you mind reviewing and merging to `2.0` and `trunk` branches only?

ewencp reviewed (approved):
Hmm, so this makes me realize we've had some divergence between converters. `ByteArrayConverter` is in the more clearly named `converters` package and is covered by the previous `*` import. I can't remember, but the `storage` package for `StringConverter` may have just been an arbitrary (and not necessarily well named) artifact of the original prototype code base.

Interestingly, JsonConverter ended up in its own dedicated namespace (though we didn't have precedence for non-primitive serializers, so that's not necessarily a divergence, and the underlying serdes are only included in Connect).

I'll merge this for now since it's fine for fixing the immediate concern, but we should also consider if we should follow up with a patch that moves the Converters into the dedicated package. We could preserve the current `StringConverter` behavior but still namespace these more conveniently for classloader isolation by replacing the current StringConverter with a trivial inheritance from the moved copy of it (though that's also maybe tricky with classpath scanning and aliases).

rhauch commented:
@ewencp good idea on relocating the new numeric converters. I've logged [KAFKA-7056](https://issues.apache.org/jira/browse/KAFKA-7056) and will do that today.
